### PR TITLE
Abi decode

### PR DESCRIPTION
**Pull request by jacky860226**

### CODE REVIEW

File: lib/CodeGen/CGExpr.cpp [hunk 1]
```diff
@@ -1689,6 +1799,24 @@ llvm::Value *CodeGenFunction::emitAbiEncode(const CallExpr *CE) {
   return Bytes;
 }
 
+ExprValuePtr CodeGenFunction::emitAbiDecode(const CallExpr *CE) {
+  // abi_encode
+  auto Arguments = CE->getArguments();
+  assert(Arguments.size() == 2);
+  auto Arg = Arguments.at(0);
+  if (auto CE = dynamic_cast<const CastExpr *>(Arg))
```
Comment (hydai) [inline lib/CodeGen/CGExpr.cpp:1807]: You should not use the same name here. Because the CE means `CallExpr` in this function arguments.

**Review by hydai — CHANGES_REQUESTED** [review]

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/CodeGen/CGExpr.cpp [hunk 2]
```diff
@@ -900,7 +900,117 @@ class AbiEmitter {
...
+
 private:
+  std::pair<std::vector<llvm::Value *>, llvm::Value *>
+  getDecodeTuple(llvm::Value *Int8Ptr, const TupleType *Ty) {
+    auto Int32Ty = IntegerType::getIntN(32);
+    const auto &ElementTypes = Ty->getElementTypes();
+    std::vector<std::pair<llvm::Value *, size_t>> DynamicPos;
+    std::vector<llvm::Value *> Vals(ElementTypes.size());
+    llvm::Value *NextInt8Ptr = Int8Ptr;
+    for (size_t i = 0; i < ElementTypes.size(); ++i) {
```
Comment (hydai) [inline lib/CodeGen/CGExpr.cpp:990]: `I`

File: lib/CodeGen/CGExpr.cpp [hunk 3]
```diff
@@ -900,7 +900,117 @@ class AbiEmitter {
...
+        ExprValuePtr ValExprValue;
+        std::tie(ValExprValue, NextInt8Ptr) =
+            getDecode(NextInt8Ptr, ElementTypes[i].get());
+        Vals[i] = ValExprValue->load(Builder, CGM);
+      }
+    }
+    for (auto &DP : DynamicPos) {
+      ExprValuePtr ValExprValue;
+      llvm::Value *Pos;
+      size_t i;
```
Comment (hydai) [inline lib/CodeGen/CGExpr.cpp:1006]: `I`

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/Sema/SemaResolveType.cpp [hunk 4]
```diff
@@ -783,14 +796,16 @@ void Sema::resolveImplicitCast(ImplicitCastExpr &IC, TypePtr DstTy,
...
     assert(DstTy->getCategory() == Type::Category::Tuple);
-    auto SrcTup = dynamic_cast<TupleExpr *>(IC.getSubExpr());
-    auto DstTupTy = dynamic_cast<const TupleType *>(DstTy.get());
-    std::size_t Num = SrcTup->getComponents().size();
-    for (std::size_t Idx = 0; Idx < Num; ++Idx) {
-      auto TIC = dynamic_cast<ImplicitCastExpr *>(SrcTup->getComponents()[Idx]);
-      if (TIC)
-        resolveImplicitCast(*TIC, DstTupTy->getElementTypes()[Idx],
-                            PrefereLValue);
+    if (auto SrcTup = dynamic_cast<TupleExpr *>(IC.getSubExpr())) {
```
[the remark this answers is not in the document]
Comment (jacky860226) [reply lib/Sema/SemaResolveType.cpp:799]: if SrcTup is nullptr then IC.getSubExpr() will be TypesTupleExpr

File: lib/Sema/SemaResolveType.cpp [hunk 5]
```diff
@@ -632,10 +661,14 @@ void TypeResolver::visit(CallExprType &CE) {
       case Identifier::SpecialIdentifier::abi_encodeWithSignature: {
         auto &Signature = CE.getRawArguments().at(0);
         if (auto *IC = dynamic_cast<ImplicitCastExpr *>(Signature.get())) {
-          Actions.resolveImplicitCast(*IC, std::make_shared<BytesType>(),
+          Actions.resolveImplicitCast(*IC, std::make_shared<StringType>(),
```
Comment (hydai) [inline lib/Sema/SemaResolveType.cpp:664]: Is the signature type `BytesType`? Why does this change from `BytesType` to `StringType`?

Comment (jacky860226) [reply lib/Sema/SemaResolveType.cpp:664]: StringType cannot ImplicitCast to BytesType in solidity, so I use ExplicitCast here.

**Review by hydai — APPROVED** [review]

**Pull request merged**